File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/1116-print-zero-even-odd.cpp
#include <thread>
#include <mutex>
#include <condition_variable>

class ZeroEvenOdd {
private:
    int n;
    int current;
    std::mutex mtx;
    std::condition_variable cv;

public:
    ZeroEvenOdd(int n) {
        this->n = n;
        this->current = 0;
    }

    // You may call this function multiple times, even from different threads.
    void zero(std::function<void(int)> printNumber) {
        for (int i = 1; i <= n; ++i) {
            std::unique_lock<std::mutex> lock(mtx);
            cv.wait(lock, [this] { return current == 0; });
            printNumber(0);
            if (i % 2 == 1) {
                current = 1;
            } else {
                current = 2;
            }
            cv.notify_all();
        }
    }

    void even(std::function<void(int)> printNumber) {
        for (int i = 2; i <= n; i += 2) {
            std::unique_lock<std::mutex> lock(mtx);
            cv.wait(lock, [this] { return current == 2; });
            printNumber(i);
            current = 0;
            cv.notify_all();
        }
    }

    void odd(std::function<void(int)> printNumber) {
        for (int i = 1; i <= n; i += 2) {
            std::unique_lock<std::mutex> lock(mtx);
            cv.wait(lock, [this] { return current == 1; });
            printNumber(i);
            current = 0;
            cv.notify_all();
        }
    }
};

File: leetcode/AIcoding/3.DeepSeek-v3/javascript_solutions/1116-print-zero-even-odd.js
class ZeroEvenOdd {
    constructor(n) {
        this.n = n;
        this.current = 1;
        this.zeroTurn = true;
        this.lock = new Promise(resolve => resolve());
    }

    async zero(printNumber) {
        for (let i = 0; i < this.n; i++) {
            await this.lock;
            if (!this.zeroTurn) continue;
            printNumber(0);
            this.zeroTurn = false;
            this.lock = new Promise(resolve => {
                this.resolve = resolove;
            });
            this.resolve();
        }
    }

    async even(printNumber) {
        while (this.current <= this.n) {
            await this.lock;
            if (this.zeroTurn || this.current % 2 !== 0) continue;
            printNumber(this.current);
            this.current++;
            this.zeroTurn = true;
            this.lock = new Promise(resolve => {
                this.resolve = resolve;
            });
            this.resolve();
        }
    }

    async odd(printNumber) {
        while (this.current <= this.n) {
            await this.lock;
            if (this.zeroTurn || this.current % 2 === 0) continue;
            printNumber(this.current);
            this.current++;
            this.zeroTurn = true;
            this.lock = new Promise(resolve => {
                this.resolve = resolve;
            });
            this.resolve();
        }
    }
}